### ISSUE TITLE
[pickers] DesktopTimePicker doesn't mask and formatting correctly on ampm 

### ISSUE THREAD

**Issue by josejimenez-swipejobs** [mui/mui-x#4680]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

The current `DesktopTimePicker` component does not handle mask and formatting when the `ampm` prop is `true` when the keyboard is been used to enter the time.

1. `DesktopTimePicker` is not accepting another `mask` prop. It seems to be that the mask is been set always as `__:__` and it is been disabled when ampm is on (https://github.com/mui/mui-x/blob/master/packages/x-date-pickers/src/TimePicker/shared.ts#L67-L68).
2. Also, there is a problem with the formatting that is allowing any value on any position on the input.

This is the CodeSandbox of the current implementation: https://codesandbox.io/s/mui-x-time-picker-example-0g8kxd?file=/src/App.js. This allows entering any value in the input.

### Expected behavior 🤔

It will be great to handle the mask on the input as well as to format the input based on the mask and format of time entered.

The previous version of the `@material-ui/pickers` version 3 handled the mask and formatting of the input.

This is a CodeSandbox of the previous implementation: https://codesandbox.io/s/material-ui-pickers-version-3-1c1qjv?file=/src/App.js

The formatting method is slide different on the material UI pickers:
https://github.com/mui/material-ui-pickers/blob/v3-x/lib/src/_shared/KeyboardDateInput.tsx#L74-L79

### Steps to reproduce 🕹

Steps:

1. It should be passed the prop `ampm` on the `DesktopTimePicker` component from the `@mui/x-date-pickers`

### Context 🔦

It is mainly to be used to manually enter the time in format ampm for a date entry website. The idea is to allow users to type the time on a big table with several inputs on it.

I have made a partial solution to the problem but making changes with the package `patch-package` on the file https://github.com/mui/mui-x/blob/master/packages/x-date-pickers/src/TimePicker/shared.ts#L67-L68, with the following patch

```patch
diff --git a/node_modules/@mui/x-date-pickers/TimePicker/shared.js b/node_modules/@mui/x-date-pickers/TimePicker/shared.js
index 8987c09..8d23856 100644
--- a/node_modules/@mui/x-date-pickers/TimePicker/shared.js
+++ b/node_modules/@mui/x-date-pickers/TimePicker/shared.js
@@ -23,8 +23,8 @@ export function useTimePickerDefaultizedProps(props, name) {
     openTo: 'hours',
     views: ['hours', 'minutes'],
     acceptRegex: ampm ? /[\dapAP]/gi : /\d/gi,
-    mask: '__:__',
-    disableMaskedInput: ampm,
+    mask: themeProps.mask,
+    disableMaskedInput: themeProps.disableMaskedInput != null ? themeProps.disableMaskedInput : false,
     getOpenDialogAriaText: getTextFieldAriaText,
     inputFormat: ampm ? utils.formats.fullTime12h : utils.formats.fullTime24h
   }, themeProps, {

```

And to fix the format, I'm passing the function from the `@material-ui/pickers` version 3 to the prop `rifmFormatter`.
Function: https://github.com/mui/material-ui-pickers/blob/554e6d4af535e75752d62aa1b2296b56feb3a86f/lib/src/_helpers/text-field-helper.ts#L140
Implementation:
```tsx
const maskSymbol: string = '_';

const makeMaskFromFormat = (format: string, numberMaskChar: string): string =>
  format.replace(/[a-z]/gi, numberMaskChar);

const inputMask: string = mask || makeMaskFromFormat(format, maskSymbol);

const refuse = /[^\dap]+/gi

const formatter: (value: string) => string = React.useMemo(
     // function from  https://github.com/mui/material-ui-pickers/blob/554e6d4af535e75752d62aa1b2296b56feb3a86f/lib/src/_helpers/text-field-helper.ts#L140
    () => maskedDateFormatter(inputMask, maskSymbol, refuse),
    [inputMask, refuse],
  );

return (
   <DesktopTimePicker
       {...props}
       ampm
       mask="__:__ _M"
       rifmFormatter={formatter}
   />
);
```

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  System:
    OS: macOS 11.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 72.17 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.1 - ~/.nvm/versions/node/v14.16.1/bin/node
    npm: 6.14.12- ~/usr/local/bin/npm
  Managers:
    Homebrew: 3.2.15 - /usr/local/bin/brew
    pip3: 19.2.3 - /usr/bin/pip3
    RubyGems: 3.0.3 - /usr/bin/gem
  Utilities:
    Make: 3.81 - /usr/bin/make
    GCC: 4.2.1 - /usr/bin/gcc
    Git: 2.30.1 - /usr/bin/git
    Clang: 12.0.5 - /usr/bin/clang
  Servers:
    Apache: 2.4.48 - /usr/sbin/apachectl
  IDEs:
    IntelliJ: 2021.2.2
    Nano: 2.0.6 - /usr/bin/nano
    Vim: 8.2 - /usr/bin/vim
    Xcode: /undefined - /usr/bin/xcodebuild
  Languages:
    Bash: 3.2.57 - /bin/bash
    Perl: 5.30.2 - /usr/bin/perl
    PHP: 7.3.29 - /usr/bin/php
    Python: 2.7.16 - /usr/bin/python
    Python3: 3.8.2 - /usr/bin/python3
    Ruby: 2.6.3 - /usr/bin/ruby
  Databases:
    SQLite: 3.32.3 - /usr/bin/sqlite3
  Browsers:
    Chrome: 101.0.4951.41
    Firefox: 99.0.1
    Safari: 14.1.2

```
</details>

**Comment by alexfauquette** [comment]
Thanks for the detailed explainations.

As a workaround, you can override default value
```jsx
        disableMaskedInput={false}
        mask="__:__ _m"
        ```
https://codesandbox.io/s/mui-x-time-picker-example-forked-vvkgk2?file=/src/App.js:658-716

Related to #4488

**Comment by josejimenez-swipejobs** [comment]
Thanks @alexfauquette, I will use that in the meanwhile

**Comment by alexfauquette** [comment]
Has been solved by #4731
